### ISSUE TITLE
The command ng serve is not responding to changes to ng module

### ISSUE THREAD

**Issue by wagnermaciel** [angular/angular-cli#16376]
# 🐞 Bug report

### Command

- [ ] new
- [ ] build
- [x] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Description

`ng serve` either is not responding to changes made to the `declarations` property of an `NgModule`.

## 🔬 Minimal Reproduction
1. `ng new pokemon --routing=false --style=css`
2. `ng generate component pokedex`
3. set up `pokedex` to be used by the root component
4. `ng serve`
5. remove `pokedex` from the `declarations` field of the `NgModule` in `app.module.ts`

### Expected
The app crashes
### Actual
The app continues working

### Additional Information
Restarting `ng serve` fixes the issue.
This also works vice-versa, so if the application is crashing due to `pokedex` _not_ being in the `declarations` of `app.module.ts`, adding it without restarting `ng serve` will still be crashing.

## 🌍 Your Environment
![image](https://user-images.githubusercontent.com/25158423/70344220-fe669a80-180d-11ea-87ef-afb74aacd87a.png)

## Demo
I've created a demo project [here](https://github.com/wagnermaciel/Pokedex).
To reproduce this issue
1. `git clone https://github.com/wagnermaciel/Pokedex.git`
2. `npm install`
3. `ng serve`
4. Go to http://localhost:4200/pikachu
4. remove `PokedexComponent` from the `declarations` of the `NgModule` in `src/app.app.module.ts`

**Comment by filipesilva** [comment]
I think that's actually related to `@angular/core` and the rest of the framework being at rc.2. I know that there were some incorrect rebuild states in that version. Can you try with the latest RC?

**Comment by wagnermaciel** [comment]
@filipesilva Sorry, I'm still a little new here. Could you clarify what you mean by "the latest RC"? Based on the output from running `ng version`, I thought that all of my angular env is using rc.5

**Comment by filipesilva** [comment]
Yes rc.5 is the latest RC, but https://github.com/wagnermaciel/Pokedex/blob/master/package.json was using rc.2 so I just wanted to make sure. Will look into it, thanks!

**Comment by wagnermaciel** [comment]
Ah, got it. I'll take another look when I get the chance. Thank you!

**Comment by wagnermaciel** [comment]
I tried this out again and it seems to be fixed - somewhere between `rc.5 and rc.8`. There is still some weirdness where deleting a necessary declaration of a component will rebuild successfully (which is the unexpected/incorrect outcome) and then after a couple seconds (correctly) fail, but this is much more minor than the issue I initially observed.

@filipesilva I spoke with @dgp1130 and we concluded that although `package.json` says rc.2 it seems like my global version is what was getting used - in the above case `rc.5` and currently `rc.8`. I believe this is because of the semver's in the angular version declarations in `package.json`.

**Comment by wagnermaciel** [comment]
Seems like I was able to reproduce this. I believe this issue is mainly to do with how `ng serve` interacts with Angular's `RouterModule`.

I've created as simple an angular app to demo the bug as I could.
https://github.com/wagnermaciel/bug-demo/tree/master

As a recap - The bug happens when a component is undeclared but used through the `RouterModule`. Incremental builds will not detect adding or removing the component from the `NgModule`'s declarations, but killing and restarting the server will.

Here is my current environment

![Screen Shot 2020-01-15 at 12 05 42 PM](https://user-images.githubusercontent.com/25158423/72467237-5be7f280-378f-11ea-93fe-50be95b80faf.png)

cc: @dgp1130

**Comment by filipesilva** [comment]
I think the answer to the `PikachuComponent`s "where are you?" question will likely be found in the Framework repository, as the CLI doesn't do much to affect the result of AOT compilation in this case. Moved it there.
